packages/client/public/service-worker.js

```diff
@@ -0,0 +1,127 @@
+// This runs as a service worker and communicates with the content script layout.js via message passing
+// See https://developer.chrome.com/docs/extensions/how-to/devtools/extend-devtools for an intro to the architecture.
+
+const agent = typeof globalThis.browser !== "undefined" ? globalThis.browser : globalThis.chrome;
+
+const connections = {};
+
+agent.runtime.onConnect.addListener((connection) => {
+	const listener = (message, sender, sendResponse) => {
+		if (message.name == "podium/init") {
+			connections[message.tabId] = connection;
+			try {
+				onInit(connection, message);
+			} catch (e) {
+				console.error(e);
+			}
+			return;
+		}
+		if (message.name == "podium/update-headers") {
+			try {
+				onUpdateHeaders(connection, message);
+			} catch (e) {
+				console.error(e);
+			}
+			return;
+		}
+	};
+
+	connection.onMessage.addListener(listener);
+	connection.onDisconnect.addListener(() => {
+		connection.onMessage.removeListener(listener);
+
+		var tabs = Object.keys(connections);
+		for (var i = 0, len = tabs.length; i < len; i++) {
+			if (connections[tabs[i]] == connection) {
+				delete connections[tabs[i]];
+				break;
+			}
+		}
+	});
+});
+
+// Receive message from content script and relay to the devTools page for the current tab
+agent.runtime.onMessage.addListener((request, sender, sendResponse) => {
+	// Messages from content scripts should have sender.tab set
+	if (sender.tab) {
+		var tabId = sender.tab.id;
+		if (tabId in connections) {
+			connections[tabId].postMessage(request);
+		} else {
+			console.warn("Tab not found in connection list.");
+		}
+	} else {
+		console.warn("sender.tab not defined.");
+	}
+	return true;
+});
+
+async function onInit(connection, message) {
+	const defaultRules = {
+		// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/Rule
+		addRules: [
+			{
+				id: 1,
+				// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/RuleAction
+				action: {
+					type: "modifyHeaders",
+					// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/ModifyHeaderInfo
+					requestHeaders: [
+						{ header: "x-podium-app-id", operation: "set", value: "lib.podium.app@1.2.3" },
+						{ header: "x-podium-base-font-size", operation: "set", value: "1rem" },
+						{ header: "x-podium-device-type", operation: "set", value: "desktop" },
+					],
+				},
+				// https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/RuleCondition
+				condition: {
+					resourceTypes: ["main_frame"],
```

Comment:
We discussed in planning today that it would be nice if these headers were set per domain. Maybe the `"podium/init"` message sent by `headers.js` can include the URL so we can use it as a [rule condition](https://developer.chrome.com/docs/extensions/reference/api/declarativeNetRequest#properties_16) here (`urlFilter` perhaps)?

The same condition has to be set in `onUpdateHeaders` on L117.

We would have to filter out the correct rules in `activeRules` on L89 as well, so the UI doesn't show the rules for another domain.

We would also need to generate and keep track of rule IDs, not hard code to 1.

Reply:
Making an issue for this, we can do it in another PR